### PR TITLE
daemon: add job status change event

### PR DESCRIPTION
**Pull request by FeelyChau**

### CODE REVIEW

File: packages/daemon/src/app/controller/base.ts
Comment (yorkie) [inline packages/daemon/src/app/controller/base.ts:36]: This event name looks weird, the log shouldn't be related to the job.

File: packages/daemon/src/service/log-manager.ts
Comment (yorkie) [inline packages/daemon/src/service/log-manager.ts:7]: I think in log-manager, we should define something more common instead of an event to be related to a job.

**Review by yorkie — CHANGES_REQUESTED** [review]

File: packages/daemon/src/service/log-manager.ts
Comment (FeelyChau) [reply packages/daemon/src/service/log-manager.ts]: Updated.

File: packages/daemon/src/app/controller/base.ts
Comment (FeelyChau) [reply packages/daemon/src/app/controller/base.ts]: Updated.

**Comment by FeelyChau** [comment]
Refactor the log-manager, may I ask you to review again please? @yorkie

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [inline packages/daemon/src/service/trace-manager.ts:18]: How about extends from EventEmitter?

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [inline packages/daemon/src/service/trace-manager.ts:35]: log might be a event

**Comment by yorkie** [comment]
@FeelyChau The current implementation looks too complicated, may I ask you to have more improvements?

**Comment by FeelyChau** [comment]
> @FeelyChau The current implementation looks too complicated, may I ask you to have more improvements?

Yep, I review the code and feel the same.
Working on it.

File: packages/daemon/src/service/trace-manager.ts
Comment (FeelyChau) [reply packages/daemon/src/service/trace-manager.ts:19]: I think it's not a good idea for parameter type checking.

**Comment by FeelyChau** [comment]
> @FeelyChau The current implementation looks too complicated, may I ask you to have more improvements?

Done.

**Comment by FeelyChau** [comment]
Bug found: #530

**Comment by FeelyChau** [comment]
> Bug found: #530

Solved. It has nothing to do with this PR. Please review again please @yorkie , thanks!

File: packages/daemon/src/service/pipeline.ts
Comment (yorkie) [inline packages/daemon/src/service/pipeline.ts:243]: Could you please wrap this function with start/end emits?

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [reply packages/daemon/src/service/trace-manager.ts:25]: Hmm, it's not related to parameter types, what I just call is to use `super()` instead of `this.emitter`.

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [inline packages/daemon/src/service/trace-manager.ts:32]: Please make a consistent style for comments.

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [inline packages/daemon/src/service/trace-manager.ts:37]: What do the above fields mean?

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [inline packages/daemon/src/service/trace-manager.ts:47]: Could you please explain a bit the relationship between logger with trace?

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [inline packages/daemon/src/service/trace-manager.ts:61]: Also, explain what's the event in a trace object.

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [inline packages/daemon/src/service/trace-manager.ts:89]: Ok, I got you but here is there necessary to use type here?

**Comment by yorkie** [comment]
Thank you @FeelyChau and please this PR tries to refactor the internal module, so we have to explain more details includes:

- the new class `Trace`.
- the improvement by this refactoring.

**Comment by yorkie** [comment]
And BTW because we will have a stable release next week, so I'm -1 on landing this refactored change, is there possible to just support emiting these events out?

**Comment by yorkie** [comment]
> And BTW because we will have a stable release next week, so I'm -1 on landing this refactored change, is there possible to just support emiting these events out?

Or we need almost full tests before landing this :)

File: packages/daemon/src/service/trace-manager.ts
Comment (FeelyChau) [reply packages/daemon/src/service/trace-manager.ts:89]: The `type` parameter is used for client to know what the data is.
So, the types of Type and data are bound, and we need to wrap them.

File: packages/daemon/src/service/trace-manager.ts
Comment (FeelyChau) [reply packages/daemon/src/service/trace-manager.ts:25]: That will expose `emit`/`on` method, IMO, the relationship between the EventEmitter and Tracer is one of inclusion rather than inheritance.
A `Tracer` object **HAS** an event channel, not **IS**, It also has the log stream.

File: packages/daemon/src/service/trace-manager.ts
Comment (FeelyChau) [reply packages/daemon/src/service/trace-manager.ts:85]: Updated, put the annotation on the class.

File: packages/daemon/src/service/trace-manager.ts
Comment (FeelyChau) [reply packages/daemon/src/service/trace-manager.ts:99]: Updated, put the annotation on the class.

File: packages/daemon/src/service/trace-manager.ts
Comment (FeelyChau) [reply packages/daemon/src/service/trace-manager.ts]: Done.

File: packages/daemon/src/service/trace-manager.ts
Comment (FeelyChau) [reply packages/daemon/src/service/trace-manager.ts:75]: Annotation added.

File: packages/daemon/src/service/trace-manager.ts
Comment (FeelyChau) [reply packages/daemon/src/service/trace-manager.ts]: Updated.

File: packages/daemon/src/service/pipeline.ts
Comment (FeelyChau) [reply packages/daemon/src/service/pipeline.ts]: Done.

**Comment by FeelyChau** [comment]
> And BTW because we will have a stable release next week, so I'm -1 on landing this refactored change, is there possible to just support emiting these events out?

Could we create version branches to avoid this problem?
We could create  branch 1.2-dev for bug fix and 1.3-dev for feature improvement.

File: packages/daemon/src/app/controller/base.ts
Comment (yorkie) [inline packages/daemon/src/app/controller/base.ts:8]: The name is too general, it's recommended to use the package name as a prefix.

File: packages/daemon/src/service/pipeline.ts
Comment (yorkie) [inline packages/daemon/src/service/pipeline.ts:227]: In the JavaScript world, a emit function is more like to declare as: `emit(name, value)`.

File: packages/daemon/src/service/pipeline.ts
Comment (yorkie) [inline packages/daemon/src/service/pipeline.ts:248]: How about `run('dataCollect', ...args)`?

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [inline packages/daemon/src/service/trace-manager.ts:9]:
```suggestion
const debug = Debug('daemon.service.tracer');
```

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [inline packages/daemon/src/service/trace-manager.ts:11]:
```suggestion
export type TraceEventType = 'log' | 'job_status';
```

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [inline packages/daemon/src/service/trace-manager.ts:16]:
```suggestion
export class TraceEvent {
```

Comment (yorkie) [reply packages/daemon/src/service/trace-manager.ts:16]: Maybe interface?

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [inline packages/daemon/src/service/trace-manager.ts:24]: Could this be defined at JobController?

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [inline packages/daemon/src/service/trace-manager.ts:37]:
```suggestion
type LogLevel = 'info' | 'warn' | 'error';
```

It's used in this file only.

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [inline packages/daemon/src/service/trace-manager.ts:54]: I think this should be more general.

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [inline packages/daemon/src/service/trace-manager.ts:56]: Ditto, shall we implement the log parts based on the event?

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [inline packages/daemon/src/service/trace-manager.ts:67]: Ditto, in a `Trace` object, maintain `stdout` and `stderr` looks weird a bit, extending or registration would be more reasonable.

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [reply packages/daemon/src/service/trace-manager.ts:85]: I have commented some, and the logger should be a module(class) based on the Trace.

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [inline packages/daemon/src/service/trace-manager.ts:94]: Ditto.

**Comment by yorkie** [comment]
@FeelyChau I think a tracer is more like a connector, which connects to a target object and dispatches some events to clients, for example, someone wants to track the plugin installing logs:

- the related logs might be some trace-able objects, which the tracer will connect to them.
- the tracer object maintains the connections to their clients(from controllers), and dispatch events to them.
- there is a trace-able object called `stdout`, which emits a `data` event, the tracer should promise this event could be sent to clients.

For JobStatusChange, it's similar to the above, we should define a trace-able object for `JobModel`, and dispatch events to clients.

**Comment by yorkie** [comment]
With the above design, the `Trace` could be extended to listen to the other objects in an easy way.

**Comment by yorkie** [comment]
Moreover, we need to add unit tests for `Trace`, too.

File: packages/daemon/src/service/trace-manager.ts
Comment (FeelyChau) [reply packages/daemon/src/service/trace-manager.ts:56]: Event could't be buffed, if that, we need to implement the event buffer.

File: packages/daemon/src/service/trace-manager.ts
Comment (FeelyChau) [reply packages/daemon/src/service/trace-manager.ts:24]: I think we should move them out finally, like the `JobResp`/`TraceResp`, because the sdk depends on them too.

File: packages/daemon/src/service/trace-manager.ts
Comment (FeelyChau) [reply packages/daemon/src/service/trace-manager.ts:85]: Sorry, I doesn't get it. What you mean `the logger should be a module(class) based on the Trace`? In the current implementation, the loggers are part of the tracer.

File: packages/daemon/src/app/controller/base.ts
Comment (FeelyChau) [reply packages/daemon/src/app/controller/base.ts:8]: How about `daemon.base-controller`?

File: packages/daemon/src/service/pipeline.ts
Comment (FeelyChau) [reply packages/daemon/src/service/pipeline.ts:227]: Or `push`?

File: packages/daemon/src/service/pipeline.ts
Comment (FeelyChau) [reply packages/daemon/src/service/pipeline.ts:248]: What's the `args`? If you mean the args of the function `startJob`, it's not correct for function `run`.

File: packages/daemon/src/service/trace-manager.ts
Comment (FeelyChau) [reply packages/daemon/src/service/trace-manager.ts]: Interface will make us to define `type` in every child class.

> Moreover, we need to add unit tests for `Trace`, too.

Tests has been added for tracer and trace manager.

**Comment by FeelyChau** [comment]
> Moreover, we need to add unit tests for `Trace`, too.

Tests has been added for tracer and trace manager.

File: packages/daemon/src/app/controller/base.ts
Comment (yorkie) [inline packages/daemon/src/app/controller/base.ts:8]: daemon.controller.base

File: packages/daemon/src/service/pipeline.ts
Comment (yorkie) [reply packages/daemon/src/service/pipeline.ts:227]: But here should use `emit`, may I ask you to tweak this?

File: packages/daemon/src/service/pipeline.ts
Comment (yorkie) [reply packages/daemon/src/service/pipeline.ts:248]:
```js
await run('dataCollect', getParams(plugins.dataCollect.params, { ... });
```

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [inline packages/daemon/src/service/trace-manager.ts:52]: 1. typo: `PipecookEvent` -> `PipcookEvent`
2. I don't think `PipcookEvent` is correct, they're used in daemon only.

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [reply packages/daemon/src/service/trace-manager.ts:67]: Why not moving this to LogEvent?

File: packages/daemon/src/service/trace-manager.ts
Comment (FeelyChau) [reply packages/daemon/src/service/trace-manager.ts:52]: SDK will use it too.

**Comment by yorkie** [comment]
@FeelyChau I still found the log-related fields are defined in Tracer class, which should be moved to LogEvent or other log-related classes.

File: packages/daemon/src/service/trace-manager.ts
Comment (FeelyChau) [reply packages/daemon/src/service/trace-manager.ts:67]: LogEvent is a event entity, how could it handle the stream?

File: packages/daemon/src/service/pipeline.ts
Comment (FeelyChau) [reply packages/daemon/src/service/pipeline.ts:227]: But the two parameters are binding, as we discussed above, we should depend on the typings to make it right.

Comment (yorkie) [reply packages/daemon/src/service/pipeline.ts:227]: Just add an event type for the first argument.

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [inline packages/daemon/src/service/trace-manager.ts:52]: Ditto, I think the PipcookEvent is not proper for this type, it's used in daemon only.

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [reply packages/daemon/src/service/trace-manager.ts:67]: Or another type or class?

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [reply packages/daemon/src/service/trace-manager.ts:86]: > the loggers are part of the tracer.

This looks weird, why does the tracer own the log-related props?

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [inline packages/daemon/src/service/trace-manager.ts:127]: We may use `dispatchEvent(event)` instead of `emit`, see https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/dispatchEvent.

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [inline packages/daemon/src/service/trace-manager.ts:223]: How about moving the stdout/stderr into this class?

File: packages/daemon/src/service/trace-manager.ts
Comment (yorkie) [inline packages/daemon/src/service/trace-manager.ts:170]:
```suggestion
  constructor(filename: string) {
```

**Review by yorkie — DISMISSED** [review]
LGTM if CI is green

**Review by yorkie — APPROVED** [review]

**Review by rickycao-qy — APPROVED** [review]
LGTM

**Review by WenheLI — APPROVED** [review]
LGTM

**Pull request merged**
